File: strategic-indicators-api/si_app/application/services/strategic_indicators/indicator_source_key_validation.py
from __future__ import annotations

import re

SOURCE_KEY_PATTERN = re.compile(r"^[a-z][a-z0-9_]{0,127}$")


def normalize_indicator_source_key(source_key: str | None) -> str | None:
    normalized = (source_key or "").strip()
    return normalized or None


def validate_indicator_source_key(
    source_key: str | None,
    *,
    is_active: bool,
) -> None:
    normalized = normalize_indicator_source_key(source_key)

    if is_active and not normalized:
        raise ValueError(
            "source_key é obrigatório para indicadores ativos "
            "(medições automáticas e metas nos dashboards departamentais)."
        )

    if not normalized:
        return

    if not SOURCE_KEY_PATTERN.match(normalized):
        raise ValueError(
            "source_key inválido: use letras minúsculas, números e underscore "
            "(ex.: commercial_rol, production_otd)."
        )

File: plugins/strategic-indicators/src/ui/utils/indicatorSourceKeyValidation.ts
const SOURCE_KEY_PATTERN = /^[a-z][a-z0-9_]{0,127}$/;

export function normalizeIndicatorSourceKey(sourceKey: string | null | undefined): string {
  return (sourceKey ?? "").trim();
}

export function validateIndicatorSourceKey(
  sourceKey: string | null | undefined,
  isActive: boolean,
): string | null {
  const normalized = normalizeIndicatorSourceKey(sourceKey);

  if (isActive && !normalized) {
    return (
      "A chave da fonte é obrigatória para indicadores ativos " +
      "(medições automáticas e metas nos dashboards departamentais)."
    );
  }

  if (!normalized) {
    return null;
  }

  if (!SOURCE_KEY_PATTERN.test(normalized)) {
    return (
      "Chave inválida: use letras minúsculas, números e underscore " +
      "(ex.: commercial_rol, production_otd)."
    );
  }

  return null;
}
